### PR TITLE
Proxy properties

### PR DESCRIPTION
**Pull request by liav-certora**

### CODE REVIEW

File: certora/confs/ProxySimulator.conf [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        "SafeWebAuthnSignerProxy:_VERIFIERS=P256",
+        "SafeWebAuthnSignerProxy:_SINGLETON=SafeWebAuthnSignerSingleton"
+    ],
+    "java_args": [
+        " -ea -Dlevel.setup.helpers=info"
+    ],
+    "msg": "sanity_with_all_default_summaries",
+    "process": "emv",
+    "prover_args": [
+        " -verifyCache -verifyTACDumps -testMode -checkRuleDigest -callTraceHardFail on"
```
Comment (nd-certora) [inline certora/confs/ProxySimulator.conf]: I think we don't need those and also other, java_args.
We need to prepare for passing to safe and will have it in their ci. we can look at other configs they have

File: certora/confs/ProxySimulator.conf [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    "packages":[
+        "@safe-global=node_modules/@safe-global",
+        "@account-abstraction=node_modules/@account-abstraction"
+    ],
+    "solc": "solc8.23",
+    "solc_via_ir": false,
+    "optimistic_loop": true,
+    "loop_iter": "6",
+    "optimistic_hashing": true,
+    "prover_version": "master",
```
Comment (nd-certora) [inline certora/confs/ProxySimulator.conf]: do we need master now?

File: modules/passkey/contracts/ProxySimulator.sol [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+/* solhint-disable no-complex-fallback */
+pragma solidity >=0.8.0;
```
Comment (nd-certora) [inline modules/passkey/contracts/ProxySimulator.sol]: move this file to certora folder under helpers

File: modules/passkey/contracts/SafeWebAuthnSignerProxyHarness.sol [hunk 4]
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: LGPL-3.0-only
```
Comment (nd-certora) [inline modules/passkey/contracts/SafeWebAuthnSignerProxyHarness.sol]: move this file to certora folder under helpers

File: modules/passkey/contracts/SafeWebAuthnSignerProxyHarness.sol [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+/* solhint-disable no-complex-fallback */
+pragma solidity >=0.8.0;
+
+import {P256} from "./libraries/WebAuthn.sol";
+import {SafeWebAuthnSignerProxy} from "./SafeWebAuthnSignerProxy.sol";
+
+/**
+ * @title Safe WebAuthn Signer Proxy
```
Comment (nd-certora) [inline modules/passkey/contracts/SafeWebAuthnSignerProxyHarness.sol]: fix the comments and explain this file, it's a copy of the fallback with the delegate, right?

File: certora/specs/ProxySimulator.spec [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    ] default NONDET;
+}
+
+/*
+Property 14. Proxy - verify return data from the fallback is only one of the magicNumbers
+Uses another contract that simulates interaction with the proxy. The reason is that the prover doesn't check all
+possible calldata values so this simualtion will make the prover choose different values that will be passed on the calldata.
+Rule stuck.
+*/
+rule proxyReturnValue {
```
Comment (nd-certora) [inline certora/specs/ProxySimulator.spec:19]: so what's the status with this rule?

File: certora/specs/ProxySimulator.spec [hunk 7]
```diff
@@ -0,0 +1,25 @@
+using SafeWebAuthnSignerProxy as SafeWebAuthnSignerProxy;
+
+definition MAGIC_VALUE() returns bytes4 = to_bytes4(0x1626ba7e);
```
Comment (nd-certora) [inline certora/specs/ProxySimulator.spec:4]: comment, what is corresponding to

File: certora/specs/SafeWebAuthnSignerProxy.spec [hunk 8]
```diff
@@ -21,10 +22,35 @@ use rule timeoutChecker filtered { f -> f.contract == currentContract }
...
+
+hook DELEGATECALL(uint g, address addr, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc {
+    // DELEGATECALL is used in this contract, but it only ever calls into the singleton.
+    assert (executingContract != currentContract || addr == SafeWebAuthnSignerSingleton,
+        "we should only `delegatecall` into the singleton."
+    );
+}
+
+/*
+Property 12. Proxy - Delegate Call Integrity (calls the Singleton)
```
Comment (nd-certora) [inline certora/specs/SafeWebAuthnSignerProxy.spec:36]: where is the list of properties, if we give numbers we should have the list someplace (properties.md or readme.md under certora folder)

File: certora/confs/ProxySimulator.conf (hunk 1 above)
Comment (liav-certora) [reply certora/confs/ProxySimulator.conf]: Agreed. removed.

File: certora/confs/ProxySimulator.conf (hunk 2 above)
Comment (liav-certora) [reply certora/confs/ProxySimulator.conf]: No actually.

File: modules/passkey/contracts/ProxySimulator.sol (hunk 3 above)
Comment (liav-certora) [reply modules/passkey/contracts/ProxySimulator.sol]: moved

File: modules/passkey/contracts/SafeWebAuthnSignerProxyHarness.sol (hunk 4 above)
Comment (liav-certora) [reply modules/passkey/contracts/SafeWebAuthnSignerProxyHarness.sol]: moved under harnesses

File: modules/passkey/contracts/SafeWebAuthnSignerProxyHarness.sol (hunk 5 above)
Comment (liav-certora) [reply modules/passkey/contracts/SafeWebAuthnSignerProxyHarness.sol]: yes, I changes the comments

File: certora/specs/ProxySimulator.spec (hunk 7 above)
Comment (liav-certora) [reply certora/specs/ProxySimulator.spec:4]: added

File: certora/specs/SafeWebAuthnSignerProxy.spec (hunk 8 above)
Comment (liav-certora) [reply certora/specs/SafeWebAuthnSignerProxy.spec:36]: Under certora folder we have `properties.txt`

File: certora/specs/ProxySimulator.spec (hunk 6 above)
Comment (liav-certora) [reply certora/specs/ProxySimulator.spec:19]: We can't make the prover call the fallback correctly. There's an open ticket on related to this issue: https://certora.atlassian.net/browse/CERT-6266

File: certora/harnesses/SafeWebAuthnSignerProxyHarness.sol [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    function fallbackButNotDelegating(uint256 x, uint256 y, P256.Verifiers verifiers) external payable returns (bytes4) {
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            // Forward the call to the singleton implementation. We append the configuration to the
+            // calldata instead of having the singleton implementation read it from storage. This is
+            // both more gas efficient and required for ERC-4337 compatibility. Note that we append
+            // the configuration fields in reverse order since the fields are packed, and this makes
+            // it so we don't need to mask any bits from the `verifiers` value. This computes `data`
+            // to be `abi.encodePacked(msg.data, x, y, verifiers)`.
```
Comment (nd-certora) [inline certora/harnesses/SafeWebAuthnSignerProxyHarness.sol:38]: this comment is now confusing, so just explain that it has only the arguments passing

**Review by nd-certora — APPROVED** [review]

**Pull request merged**